File: windows/tests/ConnectorControl.Core.Tests/TestSupport/DelayQueue.cs
namespace ConnectorControl.Core.Tests.TestSupport;

/// <summary>Captures AppHost.Delay calls so tests decide when "later" happens.</summary>
public sealed class DelayQueue
{
    public List<(TimeSpan Delay, Action Action)> Pending { get; } = [];

    public void Add(TimeSpan delay, Action action) => Pending.Add((delay, action));

    /// <summary>Runs the oldest pending action (a periodic action may re-add itself).</summary>
    public void RunNext()
    {
        var (_, action) = Pending[0];
        Pending.RemoveAt(0);
        action();
    }
}

File: Tests/ConnectorControlStateTests/TestSupport/DelayQueue.swift
import Foundation
@testable import ConnectorControlState

/// Captures AppHost.delay calls so tests decide when "later" happens.
@MainActor
final class DelayQueue {
    private(set) var pending: [(delay: TimeInterval, action: MainActorAction)] = []

    nonisolated func add(_ delay: TimeInterval, _ action: @escaping MainActorAction) {
        MainActor.assumeIsolated { pending.append((delay, action)) }
    }

    /// Runs the oldest pending action (a periodic action may re-add itself).
    func runNext() {
        let next = pending.removeFirst()
        next.action()
    }
}
